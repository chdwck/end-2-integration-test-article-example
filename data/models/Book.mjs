import knex from '../db.mjs'

class Book {
    static tableName = 'book';

    constructor({ id = null, title = null, authorName = '' }) {
       if (title === null) {
           throw new Error('param.title is required to create a Book.');
       }

       this.id = id;
       this.title = title;
       this.authorName = authorName;
    }

    static create({ title, authorName }) {
        return new Book({ title, authorName });
    }

    static fromRow (row) {
        return new Book({
            id: row.id,
            title: row.title,
            authorName: row.author_name
        });
    }

    toRow() {
        return {
            title: this.title,
            author_name: this.authorName
        };
    }

    async save() {
        try {
            const [newRow] = await knex(Book.tableName)
                .returning('*')
                .insert(this.toRow());

            return Book.fromRow(newRow);
        } catch (e) {
            console.error('Unable to save book', e)
            throw e;
        }
    }

    static async list() {
        try {
            const books = await knex(Book.tableName).select('*')

            return books.map(Book.fromRow);
        } catch (e) {
            console.error('Unable to list books', e)
            throw e;
        }
    }
}

export default Book;