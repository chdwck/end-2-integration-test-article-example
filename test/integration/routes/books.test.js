import supertest from 'supertest'
import knex from '~/data/db';
import app from '~/app'
import Book from '~/data/models/Book'

afterEach(async () => {
 // Clean up anything we created in each test
 await knex.raw('TRUNCATE TABLE book RESTART IDENTITY CASCADE');
});

afterAll(async () => {
    // ensure we close our connection
    // after the tests are completed
    knex.destroy();
});

describe('/books', () => {
    describe('GET /', () => {
        it('returns a list of books', async () => {
            // Arrange
            await Book.create({ title: 'Dracula', authorName: 'Bram Stoker'}).save();

            // Act
            const response = await supertest(app).get('/books');

            // Assert
            expect(response.status).toMatchSnapshot();
            expect(response.body).toMatchSnapshot();
        })
    })

    describe('POST /', () => {
        it('creates a book', async () => {
            // Act
            const response = await supertest(app)
                .post('/books')
                .send({ title: 'Hyperion', authorName: 'Dan Simmons' })

            // Assert
            expect(response.status).toMatchSnapshot();
            expect(response.body).toMatchSnapshot();
        })
    })
})