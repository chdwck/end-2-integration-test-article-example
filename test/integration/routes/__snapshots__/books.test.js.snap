// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/books GET / returns a list of books 1`] = `200`;

exports[`/books GET / returns a list of books 2`] = `
Object {
  "books": Array [
    Object {
      "authorName": "Bram Stoker",
      "id": 1,
      "title": "Dracula",
    },
  ],
}
`;

exports[`/books POST / creates a book 1`] = `201`;

exports[`/books POST / creates a book 2`] = `
Object {
  "authorName": "Dan Simmons",
  "id": 1,
  "title": "Hyperion",
}
`;
